src/geometry/manifold/derivation_bundle.lean

```diff
@@ -0,0 +1,165 @@
+/-
+Copyright © 2020 Nicolò Cavalleri. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Nicolò Cavalleri
+-/
+
+import geometry.manifold.algebra.smooth_functions
+import ring_theory.derivation
+
+/-!
+
+# Derivation bundle
+
+In this file we define the derivations at a point of a manifold on the algebra of smooth fuctions.
+Moreover we define the total bundle of derivations (although at the moment it has not been given a
+topology). Finally we define the differential of a function in terms of derivations.
+
+The content of this file is not meant to be regarded as an alternative definition to the current
+tangent bundle but rather as a purely algebraic theory that provides a purely algebraic definition
+of the Lie algebra for a Lie group.
+
+-/
+
+variables {𝕜 : Type*} [nondiscrete_normed_field 𝕜]
+{E : Type*} [normed_group E] [normed_space 𝕜 E]
+{H : Type*} [topological_space H] (I : model_with_corners 𝕜 E H)
+{M : Type*} [topological_space M] [charted_space H M]
+
+open_locale manifold
+
+namespace point_derivation
+
+instance smooth_functions_algebra : algebra 𝕜 C^∞⟮I, M; 𝕜⟯ := by apply_instance
+instance smooth_functions_tower : is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ := by apply_instance
+
+/-- Evaluation at a point is a ring homomorphism. -/
+def smooth_funtion.eval' (x : M) : C^∞⟮I, M; 𝕜⟯ →+* 𝕜 :=
+{ to_fun    := λ f, f x,
+  map_one'  := rfl,
+  map_mul'  := λ f g, rfl,
+  map_zero' := rfl,
+  map_add'  := λ f g, rfl }
+
+/-- The above evaluation gives rise to an algebra structure of `C^∞⟮I, M; 𝕜⟯` on `𝕜`. -/
+def algebra (x : M) : algebra C^∞⟮I, M; 𝕜⟯ 𝕜 := (smooth_funtion.eval' I x).to_algebra
+
+/-- With the above algebra structure evaluation is actually an algebra morphism. -/
+def smooth_function.eval (x : M) :
+  @alg_hom C^∞⟮I, M; 𝕜⟯ C^∞⟮I, M; 𝕜⟯ 𝕜 _ _ _ _ (point_derivation.algebra I x) :=
+{ commutes' := λ k, rfl, ..smooth_funtion.eval' I x }
+
+/-- The scalar multiplication defined above gives rise to a module structure. -/
+def module (x : M) : module C^∞⟮I, M; 𝕜⟯ 𝕜 :=
+@algebra.to_module _ _ _ _ (point_derivation.algebra I x)
+
+lemma scalar_def {x : M} {f : C^∞⟮I, M; 𝕜⟯} {k : 𝕜} :
+  @has_scalar.smul C^∞⟮I, M; 𝕜⟯ 𝕜 (point_derivation.algebra I x).to_has_scalar f k = f x * k := rfl
+
+lemma is_scalar_tower (x : M) :
+  @is_scalar_tower 𝕜 C^∞⟮I, M; 𝕜⟯ 𝕜 _ (point_derivation.algebra I x).to_has_scalar _ :=
+{ smul_assoc := λ k f h, by { simp only [scalar_def, algebra.id.smul_eq_mul, smooth_map.coe_smul,
+  pi.smul_apply, mul_assoc]} }
+
+end point_derivation
+
+/-- The derivations at a point of a manifold. Some regard this as a possible definition of the
+tangent space -/
+@[reducible] def point_derivation (x : M) :=
```

Comment:
Is `@[reducible]` needed here? (I haven't investigated, it's just that usually `reducible` is a bad idea!)

Reply:
Yes `point_derivation` is just an abbreviation not to spell out each time the instances that you cannot write as real instances because of metavariables problems.